dfastmi/io/map_file.py

```diff
@@ -0,0 +1,347 @@
+
+        return data
+
+    @staticmethod
+    def _get_start_index(var) -> int:
+        if "start_index" in var.ncattrs():
+            return var.getncattr("start_index")
+        return 0
+    
+    def read_face_variable(self, 
+                      varname: str, 
+                      time_index_from_last: Optional[int] = None,
+    ) -> np.ndarray:
+        """
+        Read the last time step of any quantity defined at faces from a D-Flow FM map-file.
+
+        Arguments
+        ---------
+        varname : str
+            Name of the netCDF variable to be read.
+        time_index_from_last : Optional[int]
+            Time step offset index from the last time step written.
+
+        Raises
+        ------
+        ValueError
+            If the data file doesn't include a 2D mesh.
+            If it cannot uniquely identify the variable to be read.
+
+        Returns
+        -------
+        numpy.ndarray
+            1D data of the requested variable. If the variable is time-dependent, 
+            the time_index_from_last is used.
+        """
+        # open file
+        with nc.Dataset(self._map_file) as rootgrp:
+            location = "face"
+
+            # find any other variable by standard_name or long_name
+            var = rootgrp.get_variables_by_attributes(
+                standard_name=varname, mesh=self.mesh2d_name, location=location
+            )
+            if len(var) == 0:
+                var = rootgrp.get_variables_by_attributes(
+                    long_name=varname, mesh=self.mesh2d_name, location=location
+                )
+            if len(var) != 1:
+                raise ValueError(
+                    'Expected one variable for "{}", but obtained {}.'.format(
+                        varname, len(var)
+                    )
+                )
+            var = var[0]
+
+            if var.get_dims()[0].isunlimited():
+                # assume that time dimension is unlimited and is the first dimension
+                # slice to obtain last time step or earlier as requested
+                if time_index_from_last is None:
+                    time_index_from_last = 0
+                data = var[-1 - time_index_from_last, :]
+                
+            else:
+                if time_index_from_last is not None:
+                    raise ValueError(
+                        'Trying to access time-independent variable "{}" with time offset {}.'.format(
+                            varname, -1 - time_index_from_last
+                        )
+                    )
+
+                data = var[...]
+
+        return data
+
+    @property
+    def mesh2d_name(self) -> str:
+        """Get the name of the mesh2d variable.
+
+        Returns
+        -------
+        str
+            String containing the name of the mesh2d variable.
+        """
+        if not self._mesh2d_name:
+            with nc.Dataset(self._map_file) as rootgrp:
```

Comment:
what is rootgrp exactly? the _get_mesh2d_variable method has the name dataset, might that be a better name?

Reply:
the with nc.Dataset(self._map_file) as rootgrp: is used 5 times in this file.
Might it be more usefull to read the mapfile once instead of 5 different times for 5 different things?

Reply:
It also feels like this is IO behaviour.

Reply:
That's a good suggestion and I initially tried that. However it did not make the code better.
The calling code of the `MapFile` then has to make sure the file is properly closed each time. Moreover, the file has to be opened in two modes: read and append. 
In the end I decided to just open and close the file within each call so that we can ensure that the class behaves correctly without adding complexion to the calling/user code.

Reply:
Renamed rootgrp to dataset

Reply:
thx for clarifying and that you already tried it!